WordPress/gutenberg
# Implement `Tabs` in editor settings

Related: #52997

## What?

Implement the new `Tabs` component in the editor settings sidebar

## Why?

Due to its more granular nature, `Tabs` is better suited to more applications than its predecessor, `TabPanel`. Implementing it here will ensure it gets all of the accessibility and other benefits, and ongoing support, of the packaged component. This move also cuts down on the amount of custom CSS needed, as we're using a component with its own styles already built in that no longer need to be duplicated.

## How?

This PR modifies `Tabs` slightly by exposing the context that the component uses to pass data to its subcomponents. The editor settings sidebar implements an internal `ComplimentaryArea`, which in turn has an internal `Slot/Fill` that we need to pass our context through.

Now that we are exposing `Tabs.Context` we can use the relevant provider to pass the context wherever it needs to go.

## Testing Instructions
1. Open the post editor
2. Ensure the editor settings sidebar on the right hand side of the screen opens and closes as expected
3. Test both tabs (Post and Block). Ensure they display the correct sidebar/controls
4. In the main editor panel, click between a block and the post title.
5. Ensure that when you click on a block, the Block sidebar is automatically selected. When you click the post title, the Post sidebar should be displayed.
6. Open a new page in the editor. Select whatever template you'd like (we're going to edit it in a minute).
7. Confirm the tabs and editor-based selection you tested on a post also work the same way here.
8. On the Page sidebar, click on the name of the currently selected template, then click "Edit Template" in the modal that appears.
9. Confirm that the sidebar correctly enters "Template mode", showing you a Template tab instead of a Page tab.
10. Confirm once again that clicking on a block activates the Block sidebar, selecting that tab automatically.

### Testing Instructions for Keyboard


1. Open a new post in the post editor.
2. Enter text into the title field (should be focused by default)
3. Press Enter/Return to move to the first paragraph block
4. Confirm that the sidebar updated to the Block tab when the block was focused
5. Press Up Arrow to return to the Title field
6. Confirm that the Post tab is automatically selected
7. Press Tab to focus the sidebar
8. Use the arrow keys to navigate between buttons and Enter/Return/Space to activate a button and view that tab

### Known Issues

- While writing the above testing steps, I discovered the `Tabs` implementation is breaking the close button on the sidebar. Activating that button via mouse or keyboard triggers a `maximum update depth` error from React. That button is provided by the `ComplementaryAreaHeader` subcomponent. I'm looking into why and how to address that.
- By default, Tabs should place the focus on the `tablist` element, and then navigate between individual `tabs` with arrow keys. Arrowing to a tab should activate it automatically. With this implementation, when tabbing over to the sidebar the focus lands directly on the first `tab` (which semantically is a `button`). You then use the arrow keys to move between tabs/buttons, but because the actual button is focused and not the `tablist`, you need to explicitly activate the button with enter/return/space. I suspect this is being caused Tabs being nested inside a `ComplementaryArea` but need to look into it further.
